src/System/Hatrace/Signals.hs

```diff
@@ -49,3 +62,44 @@ signalMap =
     , (Signals.cpuTimeLimitExceeded, "cpuTimeLimitExceeded", "XCPU")
     , (Signals.fileSizeLimitExceeded, "fileSizeLimitExceeded", "XFSZ")
     ]
+
+sizeOfCSigset :: Int
+sizeOfCSigset = sizeof_sigset_t
+
+allSignals :: [Signals.Signal]
+allSignals = List.delete Signals.nullSignal $ Map.keys signalMap
+
+inSignalSet :: Signals.Signal -> ForeignPtr CSigset -> IO Bool
+inSignalSet sig ptr = do
+  withForeignPtr ptr $ \p -> do
+    r <- throwErrnoIfMinus1 "inSignalSet" (c_sigismember p sig)
+    return (r /= 0)
+
+emptySignalSet :: IO (ForeignPtr CSigset)
+emptySignalSet = do
+  fp <- mallocForeignPtrBytes sizeof_sigset_t
+  throwErrnoIfMinus1_ "emptySignalSet" (withForeignPtr fp $ c_sigemptyset)
+  return fp
+
+fullSignalSet :: IO (ForeignPtr CSigset)
+fullSignalSet = do
+  fp <- mallocForeignPtrBytes sizeof_sigset_t
+  throwErrnoIfMinus1_ "fullSignalSet" (withForeignPtr fp $ c_sigfillset)
+  return fp
+
+addSignal :: Signals.Signal -> ForeignPtr CSigset -> IO ()
+addSignal sig fPtr = do
+  withForeignPtr fPtr $ \ptr ->
+    throwErrnoIfMinus1_ "addSignal" (c_sigaddset ptr sig)
+
+foreign import capi unsafe "signal.h sigismember"
+  c_sigismember :: Ptr CSigset -> CInt -> IO CInt
+
+foreign import capi unsafe "signal.h sigemptyset"
+  c_sigemptyset :: Ptr CSigset -> IO CInt
+
+foreign import capi unsafe "signal.h sigfillset"
+  c_sigfillset :: Ptr CSigset -> IO CInt
+
+foreign import capi unsafe "signal.h sigaddset"
+  c_sigaddset :: Ptr CSigset -> CInt -> IO CInt
```

Comment:
For our record, the use of CAPI broke `stack repl`, with the following (which makes sense):

```
Error: bytecode compiler can't handle some foreign calling conventions
  Workaround: use -fobject-code, or compile this module to .o separately.

<no location info>: error:
    Could not find module ‘System.Hatrace’
    It is not a module in the current program, or in any known package.
```